### PR TITLE
Standalone CMake outside with TriBITS compatibility

### PR DESCRIPTION
**Pull request by jjwilke**

### CODE REVIEW

**Comment by ndellingwood** [comment]
@jjwilke 
I have a bit of feedback after hacking together a serial build with BLAS tpls enabled based on a local build of BLAS (and LAPACK) rather than relying solely on some pre-install on the system path. 
Work on TPL system changes in kokkos/kokkos#2467 that may resolve all this, but I wanted to document for reference in case changes in kokkos-kernels or this PR are necessary.

This was on tested in a serial build on my mac with gcc/6.5

Reproducer instructions and issues:

1. Configure and install kokkos library:
```
mkdir kokkos/build
cd kokkos/build

cmake -DKokkos_ENABLE_SERIAL=ON -DCMAKE_INSTALL_PREFIX=`pwd`/install ..
make install
```

2. Configure and build kokkos-kernels (attempt 1):
```
mkdir kokkos-kernels/build/serial-userblas
cd kokkos-kernels/build/serial-userblas

# Very hacky cmake line
cmake -DKokkos_DIR=<PATH_TO_KOKKOS>/kokkos/build/install/lib/cmake/Kokkos -DCMAKE_INSTALL_PREFIX=`pwd`/install -DKokkosKernels_ENABLE_TESTS=ON -DKokkosKernels_ENABLE_TPL_BLAS=ON -DBLAS_LIBRARIES="<PATH_TO_BLAS_INSTALL>/libblas.a" -DBLA_VENDOR="Generic" ../..
```

Link errors due to undefined symbols occurred like below that were resolved by adding this to the cmake configure line `" -DCMAKE_EXE_LINKER_FLAGS="-lgfortran -lm"`

Sample:
```
Undefined symbols for architecture x86_64:
  "__gfortran_st_write", referenced from:
      _xerbla_ in libblas.a(xerbla.o)
...
```

More link errors occurred after blowing away the build and reconfiguring due to undefined symbols for `?gesv` routines that are LAPACK routines and not BLAS routines - this seems like a problem to me in the TPL layer of kokkos-kernels where we should separate these routines into `KokkosLapack_Host_tpl.*` analogues to `KokkosBlas_Host_tpl.*` and add an option to enable lapack as a tpl in the build system, but I'm not certain.

I hacked a couple lines into the `kokkoskernels_tpls.cmake` to find and link lapack (added in separate review comments for reference) and then was able to links and compile with this configure line:

`cmake -DKokkos_DIR=<PATH_TO_KOKKOS>/kokkos/build/install/lib/cmake/Kokkos -DCMAKE_INSTALL_PREFIX=`pwd`/install -DKokkosKernels_ENABLE_TESTS=ON -DKokkosKernels_ENABLE_TPL_BLAS=ON -DBLAS_LIBRARIES="<PATH_TO_BLAS_INSTALL>/libblas.a" -DLAPACK_LIBRARIES="<PATH_TO_LAPACK_INSTALL>/liblapack.a" -DBLA_VENDOR="Generic" -DCMAKE_EXE_LINKER_FLAGS="-lgfortran -lm" ../..`

File: cmake/kokkoskernels_tpls.cmake [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  ELSE()
+    #bummer, this works the "hacky" way
+    TARGET_LINK_LIBRARIES(${LIBRARY} ${ARGN})
+  ENDIF()
+ENDFUNCTION(TARGET_LINK_FLAGS_PORTABLE)
+
+FUNCTION(KOKKOSKERNELS_LINK_TPLS LIBRARY)
+  IF (KOKKOSKERNELS_ENABLE_TPL_BLAS)
+    TARGET_LINK_LIBRARIES(${LIBRARY} PUBLIC ${BLAS_LIBRARIES})
+    TARGET_LINK_FLAGS_PORTABLE(${LIBRARY} PUBLIC ${BLAS_LINK_FLAGS})
```
Comment (ndellingwood) [inline cmake/kokkoskernels_tpls.cmake]: I had to add the lines below to get tests to build with BLAS TPL enabled, due to missing symbols for `?gesv` LAPACK routines causing link errors:

```
    TARGET_LINK_LIBRARIES(${LIBRARY} PUBLIC ${LAPACK_LIBRARIES})
    TARGET_LINK_FLAGS_PORTABLE(${LIBRARY} PUBLIC ${LAPACK_LINK_FLAGS})
```

File: cmake/kokkoskernels_tpls.cmake [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+IF (KOKKOSKERNELS_ENABLE_TPL_BLAS OR KOKKOSKERNELS_ENABLE_TPL_MKL)
+  ENABLE_LANGUAGE(C)
+  ENABLE_LANGUAGE(Fortran)
+  INCLUDE(FortranCInterface)
+  SET(F77_BLAS_MANGLE "(name,NAME) ${FortranCInterface_GLOBAL_PREFIX}name ## ${FortranCInterface_GLOBAL_SUFFIX}")
+ENDIF()
+
+IF (KOKKOSKERNELS_ENABLE_TPL_BLAS)
+  IF (NOT KOKKOSKERNELS_HAS_TRILINOS)
+    FIND_PACKAGE(BLAS REQUIRED)
```
Comment (ndellingwood) [inline cmake/kokkoskernels_tpls.cmake]: I had to add the lines below to get tests to build with BLAS TPL enabled, due to missing symbols for `?gesv` LAPACK routines causing link errors:

`    FIND_PACKAGE(LAPACK REQUIRED)`

**Comment by ndellingwood** [comment]
Thanks for TPL updates, all works as expected with user-built blas and lapack :)

File: cmake/kokkoskernels_tpls.cmake [hunk 3]
```diff
@@ -4,6 +4,7 @@ ENDFUNCTION()
 
 MACRO(KOKKOSKERNELS_ADD_TPL_OPTION NAME DEFAULT_VALUE DOCSTRING)
   KOKKOSKERNELS_ADD_OPTION(ENABLE_TPL_${NAME} ${DEFAULT_VALUE} BOOL ${DOCSTRING})
+  KOKKOSKERNELS_ADD_OPTION(${NAME}_DIR "" PATH "Location of ${NAME} instal root")
```
Comment (mhoemmen) [inline cmake/kokkoskernels_tpls.cmake]: "install" (spelling)

**Comment by jjwilke** [comment]
Close in favor of PR #491

**Pull request closed without merging**